src/Expander.hs

```diff
@@ -1165,25 +1165,29 @@ problemCategory (TypePatternDest {}) = TypePatternCaseCat
 requireDeclarationCat :: Syntax -> MacroDest -> Expand (DeclTreePtr, DeclOutputScopesPtr)
 requireDeclarationCat _ (DeclTreeDest dest outScopesDest) = return (dest, outScopesDest)
 requireDeclarationCat stx other =
-  throwError $ WrongSyntacticCategory stx DeclarationCat (problemCategory other)
+  throwError $
+  WrongSyntacticCategory stx (actual DeclarationCat) (expected $ problemCategory other)
```

Comment:
absolutely! That sounds good to me.

Reply:
❌ This is the only blocker

Reply:
issued a fix in 028928fa490222deb657465a877d9dc2b9cc6bd4 but I want to write the small Klister program before merging.

Reply:
Here's a small Klister program which triggers the problem: `define` only works in a declaration context (such as the top-level) but here it is called inside `example`, which expects an expression.

```
#lang "prelude.kl"

(example (define x 5))
```

When I ran this program on the `main` branch, I was surprised to see that the error message is actually correct, meaning that the arguments are _not_ swapped after all!

```
$ cabal run klister -- run toy.kl 
#[toy.kl:3.10-3.22]<(define x 5)>:
  Used in a position expecting an expression
  but is valid in a position expecting a top-level declaration or example
ExitFailure 1
```

I think what is going on is that the error message is written from the perspective of the macro, rather than from the perspective of the syntactic category. That is, the same information could have been conveyed by the following error message, in which it is the expression syntactic category which is the actual value and the declaration syntactic category which is the expected value:

```
expected: an expression
found: declaration #[toy.kl:3.10-3.22]<(define x 5)>
```

Thus, I think that the `Actual` and `Expected` newtype are actually making things worse for this particular function, as it makes the correct code look wrong and the wrong code look correct :upside_down_face: 

I still think the `Actual` and `Expected` newtypes themselves are useful, but for other functions such as `mismatch`.

For `WrongSyntacticCategory`, perhaps it would make more sense to use newtypes with names like `DestinationSyntacticCategory` and `MacroSyntacticCategory`?

Reply:
Alright let's take stock of the path we've traveled:
1) We agree that a `Actual` and `Expected` newtypes are useful and should be included in the PR.
2) But, in contrast to the prior suspicion `requireDeclarationCat` is not actually buggy because the error message of `toy.kl` is correct.
3) So the tasks are to (a) revert the commit that swaps `requireDeclarationCat` and (b) add more newtypes that give a proper denotation in the domain, suggestions are `DestinationSyntacticCategory` and `MacroSyntacticCategory`

But I'm still stuck on (2), here is why. If we look at the code that generates that error message on `main` we find:

```hs
...
  pp env (WrongSyntacticCategory stx actual expected) =
    hang 2 $ group $
    vsep [ pp env stx <> text ":"
...
```

which indicates to me that argument 2 and argument 3 to `WrongSyntacticCategory` are indeed interpreted as `Actual` and `Expected` respectively. So then if we cross reference with `requireDeclarationCat` on `main` we get:

```
requireDeclarationCat :: Syntax -> MacroDest -> Expand (DeclTreePtr, DeclOutputScopesPtr)
requireDeclarationCat _ (DeclTreeDest dest outScopesDest) = return (dest, outScopesDest)
requireDeclarationCat stx other =
  throwError $ WrongSyntacticCategory stx DeclarationCat (problemCategory other)
```

which does indeed set `DeclarationCat` to the actual value and `problemCategory other` to the expected, which _does_ still seem wrong to me _even_ if the error message was correct. Is it possible that `requireDeclarationCat` was not even called? Or I wonder if we could verify the call stack to be sure. What is surprising to me is that if I run `toy.kl` on my branch then I still get:
```
20:15:10 ❯ ./result/bin/klister run examples/failing-examples/toy.kl 
#[toy.kl:5.10-5.22]<(define x 5)>:
  Used in a position expecting a top-level declaration or example
  but is valid in a position expecting an expression
ExitFailure 1
```
That is, the error message is the same, which implies to me that `requireDeclarationCat` is not in the code execution path because we should expect to see a change after I swapped the arguments, but didn't.

But besides that hypothesized bug, I'm not in favor of `DestinationSyntacticCategory` and `MacroSyntacticCategory` because it implies `Destination` as an opposite to `Macro`. To make clear what I mean I'd rather have `Destination SyntacticCategory` and `Source SyntacticCategory` but I am unsure if there are other `Source`s than `Macro`.

Reply:
> on my branch then I still get:
> 
>     20:15:10 ❯ ./result/bin/klister run examples/failing-examples/toy.kl 
>     #[toy.kl:5.10-5.22]<(define x 5)>:
>       Used in a position expecting a top-level declaration or example
>       but is valid in a position expecting an expression
>     ExitFailure 1
> 
> That is, the error message is the same

Look closer! :)

The error message above is _not_ the same error message as the one which the `main` branch gives:

    $ cabal run klister -- run toy.kl 
    #[toy.kl:3.10-3.22]<(define x 5)>:
      Used in a position expecting an expression
      but is valid in a position expecting a top-level declaration or example
    ExitFailure 1

It is the error message on `main` which is correct, because `define` is a macro which is valid in a position expecting a top-level declaration or an example (that is, in the Declaration syntactic category), but it is used inside `(example ...)`, which expects an expression.

You are right that `pp` uses the order (actual, expected), while `requireDeclarationCat` uses the order (expected, actual). Surprisingly, both are correct, they just have a different interpretation of what "actual" and "expected" mean! `pp` is taking the point of view of the `(example ...)` context: that context _expects_ the `...` to be an expression, but the _actual_ `...` is `define`, a definition. `requireDeclarationCat` takes the point of view of the `define` macro: that macro _expects_ to be called in a definition context, but the _actual_ context is `(example ...)`, which is an expression context. That's why I would prefer something like `Context` and `Macro` instead of `Actual` and `Expected`.

> But besides that hypothesized bug, I'm not in favor of `DestinationSyntacticCategory` and `MacroSyntacticCategory` because it implies `Destination` as an opposite to `Macro`. To make clear what I mean I'd rather have `Destination SyntacticCategory` and `Source SyntacticCategory` but I am unsure if there are other `Source`s than `Macro`.

I'm not too happy with the names `DestinationSyntacticCategory` and `MacroSyntacticCategory` either, thanks for suggesting something better! I like how `Source` and `Destination` are very general, like `Actual` and `Expected`. I'm not sure these are quite the right terms here though, because we aren't talking about a path which goes from a source to a destination. In this case, the word "destination" describe a hole, or a location, which gets filled by a value. Maybe `Hole` and `Plug`? `Caller` and `Callee`? `Parameter` and `Argument`?

Reply:
Maybe `Socket` and `Plug`?

Reply:
Yes! That sounds perfect. I liked hole and fill because it fits the construction or carpentry motif (stuck, adhesive and bind)

Reply:
[Mortise and Tenon](https://en.m.wikipedia.org/wiki/Mortise_and_tenon), then?